### PR TITLE
Fix CMakeLists.txt

### PR DESCRIPTION
**Pull request by madmann91**
- Use modern CMake target_link_libraries for SDL2
- Fix bug where gitsha1.c was never compiled

### CODE REVIEW

File: src/utils/gitsha1.c.in [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 // Autogenerated - Do not edit
 
-#include "gitsha1.h"
```
Comment (vkoskiv) [inline src/utils/gitsha1.c.in:11]: This should be an absolute path, right? Now my compiler looks for `src/utils/utils/gitsha1.h`, which doesn't exist.

File: CMakeLists.txt [hunk 2]
```diff
@@ -85,28 +89,11 @@ if (NOT NO_SDL2)
 
 	if (SDL2_FOUND)
 		message(STATUS "SDL2 found! Enabling UI components.")
-		include_directories(${SDL2_INCLUDE_DIR})
```
Comment (vkoskiv) [inline CMakeLists.txt:88]: At least on my system, with SDL2 installed in `/usr/local/lib`, it fails to link SDL2 when building:

```
ld: library not found for -lSDL2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/c-ray] Error 1
make[1]: *** [CMakeFiles/c-ray.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by vkoskiv** [comment]
I forgot to mention - In addition to the cmake build system, I also use Xcode. (the .xcodeproj file in the tree is for this)
I've configured it to assume relative paths, so your change to `githsha1.c.in` there breaks that assumption.

**Comment by madmann91** [comment]
Using relative paths in generated files (that live in the build directory) will break any out-of-source build since the paths will not be the same as the source directory. Therefore, I have added the `src` folder in the include directories of the project, which handles that. The cleanest solution to allow out-of-source builds is to use relative directories with respect to the `src` base directory, or no relative path at all if the file is in the same directory (e.g. it would be fine to write `#include "gitsha1.h"` if it was from the same directory where `gitsha1.h` lives). An alternative to that is to add all possible subfolders of `src` in the include directories of the project, which I do not recommend as it would potentially introduce conflicts if you have files with the same name in different directories.

**Comment by madmann91** [comment]
Note that in your case if you only need the `utils` directory in the include path then you can just add it and change back the `#include "utils/gitsha1.h"` into a `#include "utils/gitsha1.h"`. In any case CMake should take care of that when generating XCode project files, so I don't understand why you have that problem to start with.

File: CMakeLists.txt [hunk 3]
```diff
@@ -85,28 +89,12 @@ if (NOT NO_SDL2)
...
-
-		if (MSVC)
-  			target_link_libraries(${PROJECT_NAME} ${SDL2_LIBRARY})
-		else()
-  			target_link_libraries(${PROJECT_NAME} -lpthread -lm ${SDL2_LIBRARY})
-		endif()
-		add_definitions(-DCRAY_SDL_ENABLED)
```
Comment (vkoskiv) [inline CMakeLists.txt:95]: This definition is needed for C-ray to use SDL. Unless CMake has a more standard way of telling the preprocessor that a dependency is indeed available?

**Comment by vkoskiv** [comment]
I'm not using an `.xcodeproj` generated by `cmake`, I'm using one that I maintain myself and include in the source tree.
I did try autogenerating it with `cmake` a long time ago, but the resulting project was really poorly structured, and I didn't look into it beyond that at the time.
What's the motivation for deviating from relative paths in the case of `gitsha1.h`?

File: CMakeLists.txt (hunk 3 above)
Comment (madmann91) [reply CMakeLists.txt:95]: Good catch, I will add that back.

**Comment by madmann91** [comment]
Well there are in general two robust ways to give include paths, which can be combined if required:

- Using the full path relative to the project include directory (e.g. if the source/include directories of the project are `proj/src` and `proj/include` then to include the file `proj/include/math/analysis/complex.h` one would type `#include "math/analysis/complex.h"`, regardless of the position of the including file),
- Using the name of the file to include directly _if_ the _included_ file is in the same directory as the _including_ file. This is not as robust, as things might behave in a strange way if the file is moved into another directory (this is currently what you do).

The problem is that, when including `gitsha1.h` from `gitsha1.c`, the paths are not guaranteed to be the same. Indeed, for an out-of-source build, `gitsha1.c` is in `build/src/utils` and therefore the include with just the path `gitsha1.h` doesn't work. For it to work, it is necessary to do either of those things:

- Change the include path to the first strategy for that file only, which is what this commit is doing, or
- Add the include directory `src/utils` to the include path. That is not as good, as it will mean that it becomes possible to include files that are in `utils/` without the `utils/` prefix, from anywhere in the source tree.

**Pull request merged**